packages/firestore/test/integration/remote/stream.test.ts

```diff
@@ -0,0 +1,231 @@
+  let queue: AsyncQueue;
+  let streamListener: StreamStatusListener;
+
+  beforeEach(() => {
+    queue = new AsyncQueue();
+    streamListener = new StreamStatusListener();
+  });
+
+  afterEach(() => {
+    streamListener.verifyNoPendingCallbacks();
+  });
+
+  function initializeWatchStream(): Promise<PersistentListenStream> {
+    const databaseInfo = getDefaultDatabaseInfo();
+
+    return PlatformSupport.getPlatform()
+      .loadConnection(databaseInfo)
+      .then(conn => {
+        const serializer = PlatformSupport.getPlatform().newSerializer(
+          databaseInfo.databaseId
+        );
+        return new Datastore(
+          databaseInfo,
+          queue,
+          conn,
+          new EmptyCredentialsProvider(),
+          serializer
+        );
+      })
+      .then(ds => {
+        return ds.newPersistentWatchStream(streamListener);
+      });
+  }
+
+  asyncIt('can be stopped before handshake', () => {
+    let watchStream: PersistentListenStream;
+
+    return initializeWatchStream()
+      .then(ws => {
+        watchStream = ws;
+        watchStream.start();
+        return streamListener.awaitCallback('onOpen');
+      })
+      .then(() => {
+        watchStream.stop();
+      });
+  });
+});
+
+describe('Write Stream', () => {
+  let queue: AsyncQueue;
+  let streamListener: StreamStatusListener;
+
+  beforeEach(() => {
+    queue = new AsyncQueue();
+    streamListener = new StreamStatusListener();
+  });
+
+  afterEach(() => {
+    streamListener.verifyNoPendingCallbacks();
+  });
+
+  function initializeWriteStream(): Promise<PersistentWriteStream> {
+    const databaseInfo = getDefaultDatabaseInfo();
+
+    return PlatformSupport.getPlatform()
+      .loadConnection(databaseInfo)
+      .then(conn => {
+        const serializer = PlatformSupport.getPlatform().newSerializer(
+          databaseInfo.databaseId
+        );
+        return new Datastore(
+          databaseInfo,
+          queue,
+          conn,
+          new EmptyCredentialsProvider(),
+          serializer
+        );
+      })
+      .then(ds => {
+        return ds.newPersistentWriteStream(streamListener);
+      });
+  }
+
+  asyncIt('can be stopped before handshake', () => {
+    let writeStream: PersistentWriteStream;
+
+    return initializeWriteStream()
+      .then(ws => {
+        writeStream = ws;
+        writeStream.start();
+        return streamListener.awaitCallback('onOpen');
+      })
+      .then(() => {
+        writeStream.stop();
+      });
+  });
+
+  asyncIt('can be stopped after handshake', () => {
+    let writeStream: PersistentWriteStream;
+
+    return initializeWriteStream()
+      .then(ws => {
+        writeStream = ws;
+        writeStream.start();
+        return streamListener.awaitCallback('onOpen');
+      })
+      .then(() => {
+        writeStream.writeHandshake();
+        return streamListener.awaitCallback('onHandshakeComplete');
+      })
+      .then(() => {
+        writeStream.stop();
+      });
```

Comment:
Why does the web test not verify that attempting to write before handshake throws but works after?  If we're going to the effort of porting the tests, we should make sure they're at parity (or at least add a comment for why they're not). :-)

Reply:
I totally forgot to port this part :(